sopel/irc/__init__.py

```diff
@@ -30,83 +30,91 @@
 from sopel.tools import identifiers
-from .backends import AsynchatBackend, SSLAsynchatBackend
+from .abstract_backends import AbstractIRCBackend
+from .backends import AsyncioBackend
 from .isupport import ISupport
-from .utils import CapReq, safe
+from .utils import CapReq, MyInfo, safe
+
+
+if TYPE_CHECKING:
+    from sopel.config import Config
 
 
 __all__ = ['abstract_backends', 'backends', 'utils']
 
 LOGGER = logging.getLogger(__name__)
+ERR_BACKEND_NOT_INITIALIZED = 'Backend not initialized; is the bot running?'
 
 
 class AbstractBot(abc.ABC):
     """Abstract definition of Sopel's interface."""
-    def __init__(self, settings):
+    def __init__(self, settings: Config):
         # private properties: access as read-only properties
-        self._user = settings.core.user
-        self._name = settings.core.name
+        self._user: str = settings.core.user
+        self._name: str = settings.core.name
         self._isupport = ISupport()
-        self._myinfo = None
-        self._nick = self.make_identifier(settings.core.nick)
+        self._myinfo: Optional[MyInfo] = None
+        self._nick: identifiers.Identifier = self.make_identifier(
+            settings.core.nick)
 
-        self.backend = None
+        self.backend: Optional[AbstractIRCBackend] = None
         """IRC Connection Backend."""
         self.connection_registered = False
         """Is the IRC Connection registered yet?"""
         self.settings = settings
         """Bot settings."""
-        self.enabled_capabilities = set()
+        self.enabled_capabilities: Set[str] = set()
         """A set containing the IRCv3 capabilities that the bot has enabled."""
-        self._cap_reqs = dict()
+        self._cap_reqs: Dict[str, List[CapReq]] = dict()
         """A dictionary of capability names to a list of requests."""
 
         # internal machinery
         self.sending = threading.RLock()
-        self.last_error_timestamp = None
+        self.last_error_timestamp: Optional[datetime] = None
         self.error_count = 0
-        self.stack = {}
+        self.stack: Dict[identifiers.Identifier, Dict[str, Any]] = {}
         self.hasquit = False
+        self.wantsrestart = False
         self.last_raw_line = ''  # last raw line received
 
     @property
-    def nick(self):
+    def nick(self) -> identifiers.Identifier:
         """Sopel's current nick.
 
         Changing this while Sopel is running is unsupported.
         """
         return self._nick
 
     @property
-    def user(self):
+    def user(self) -> str:
         """Sopel's user/ident."""
         return self._user
 
     @property
-    def name(self):
+    def name(self) -> str:
         """Sopel's "real name", as used for WHOIS responses."""
         return self._name
 
     @property
-    def config(self):
+    def config(self) -> Config:
         """The :class:`sopel.config.Config` for the current Sopel instance."""
         # TODO: Deprecate config, replaced by settings
```

Comment:
Oh! I remember that.

The issue is that having using the word `config` for a variable name make it more difficult to work with the import of `from sopel import config`, and I tend to make a distinction in code between "config" (i.e. the model, the concept, the specification of the configuration file) and "settings" (the actual values, the concrete object), as in `settings = config.Config(...)`.

Having `bot.settings` instead of `bot.config` makes it less confusing to me. And yes, btw, I'm still mad at `sopel.bot` and `sopel.trigger` for that exact same reason...